pkg/networkservice/common/heal/client.go

```diff
@@ -88,25 +90,72 @@ func (u *healClient) Request(ctx context.Context, request *networkservice.Networ
 
 	connInfo, loaded := u.conns.LoadOrStore(conn.GetId(), &connectionInfo{
 		conn: conn.Clone(),
-		mut:  sync.Mutex{},
+		cond: sync.NewCond(new(sync.Mutex)),
 	})
 	u.replaceConnectionPath(conn, connInfo)
 
-	conn, err := next.Client(ctx).Request(ctx, request, opts...)
-	if err != nil {
-		if !loaded {
+	defer func() {
+		if !loaded && err != nil {
 			u.conns.Delete(request.GetConnection().GetId())
 		}
+	}()
+
+	conn, err = next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
 		return nil, err
 	}
 
-	connInfo.mut.Lock()
-	defer connInfo.mut.Unlock()
+	condCh, cancel := u.condCh(connInfo)
+	defer cancel()
+
+	// Problem:
+	//   gRPC requests and streaming requests have different requirements for the underlying gRPC connection. There can
+	//   be a case on the closing gRPC connection, when it is still sending requests but fails to keep or start
+	//   a streaming request. In such case we can get a situation when we have a Connection with no monitor stream, so
+	//   it would never be healed.
+	// Solution:
+	//   We should check that monitor stream is ready for the Connection, before returning it back. This check should
+	//   wait for some time after the next.Request(), because for the [NSMgr -> Endpoint] case response may come faster
+	//   than monitor stream update.
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-condCh:
+	case <-clock.FromContext(ctx).After(100 * time.Millisecond):
```

Comment:
What was the thinking behind the choice of 100 ms here?

Reply:
100 ms is a common value for the unit testing.
It can be less or more, probably we will have more knowledge about common time for the request/response after the scalability testing.

Reply:
Any reason not to simply use the ctx.Done()  and wrap ctx.Err() it with your error message?

Reply:
If connection is dying, Request can be success for it but monitor fails to start. In such case we need to close this Connection --> close gRPC connection and try again. Waiting just on context wouldn't help here.

Reply:
I think 100ms would produce problems in the future. So this value could be less or high for different systems. I would be suggesting to not use 100ms and look at something smarter from grpc box.

Reply:
Another possibility might be a heuristic around a percentage of the time between now and the context deadline...

Reply:
So actually we can wait here additionally on `<-u.ctx.Done()` instead of timeout, because `u.ctx` will live no longer than underlying gRPC connection.